proposals/build-provenance-and-code-signing-for-homebrew.md

```diff
@@ -0,0 +1,112 @@
+3. The signing step is responsible for uploading the resulting signature to the same object storage as the bottle, and ensuring that the uploaded signature is addressable in a predictable and publicly derivable manner from just the formula and bottle's own metadata, [consistent with SLSA's guidenlines on distributing provenance](https://slsa.dev/spec/v1.0/distributing-provenance). In other words: a client that knows how to access the bottle's resource should also be able to access the bottle's signature resource with no additional metadata, similar to how [signature location and management](https://docs.sigstore.dev/cosign/signing_with_containers/#signature-location-and-management) works for signing container images with Sigstore.
+
+At the end of this flow, both the bottle and its attestation with signature should appear in the tap's associated object storage.
+
+In addition to signing for bottles as they're produced, the signing step may be used to "backfill" signatures (and corresponding attestations) for pre-existing bottles. These signatures would effectively act as notarizations for pre-existing bottles, rather than guarantees of build provenance. This "backfill" stage is expected to produce an initial attestation and signature for every pre-existing bottle, after which it will be removed and supplanted by signing during the bottle-building workflows. Removal of the "backfill" functionality after the initial backfill operation ensures that all subsequent attestations have corresponding build provenances.
+
+Once a tap has been fully signed for (i.e., all referenced bottles have either a "backfill" or build attestation), the tap must establish a piece of "flag day" metadata: the tap should designate a timestamp `T`, after which all bottles defined by formulae in the tap are signed. This fundamentally requires that the tap itself is honest about its flag day: an attacker who can replace the tap's flag day can perform a downgrade by disabling signature verification. This is no different in scope from Homebrew's existing threat model, where an attacker must not be allowed to modify files in a tap. In other words: this signing scheme only protects the bottles themselves, not the tap's repository; the tap's repository must continue to be protected as a matter of policy.
+
+### Sigstore signatures for third-party taps
+
+While not in scope for this proposal, this work would also enable signing for third-party taps, such as those maintained by independent open source organizations or companies.
+
+For third-party taps hosted on GitHub, the signing flow would be identical: any bottle-building workflow (including reusing the official one) would sign for bottles.
```

Comment:
Is it explained in verification below how you would find signatures for first- vs third-party packages?

Reply:
I think that's covered by the language above:

> The signing step is responsible for uploading the resulting signature to the same object storage as the bottle, and ensuring that the uploaded signature is addressable in a predictable and publicly derivable manner from just the formula and bottle's own metadata

In other words: when you do `brew install foo` with Homebrew, `brew` is responsible for consulting the tap metadata corresponding to `foo` and determining the appropriate download strategy for that tap.